Missing values for confounds result in incorrect number of volumes

Occasionally, confounds have missing values

For example, for `SherlockMerlin` sub-28 `MerlinMovie` task, the first volume for `FrameWiseDisplacement` is `n/a`. 

Regardless of the reason, we should be prepared to deal with such values.

Currently, the `n/a` value will be skipped (by pybids), and result in an incorrect number of volumes in the confounds file passed to nistats. Resulting in this error: 

```
  File "/opt/conda/envs/neuro/lib/python3.6/site-packages/nistats/design_matrix.py", line 351, in make_design_matrix
    'time-frames: %d' % (add_regs.shape[0], np.size(frame_times)))
AssertionError: Incorrect specification of additional regressors: length of regressors provided: 1026, number of time-frames: 1027
```

The "proper" BIDS model solution would be the replace n/as with the mean value using  a transformation, but that seems too cumbersome. Probably best if fitlins does this if any n/as are detected in confounds.

--- a/Dockerfile
+++ b/Dockerfile
@@ -81,7 +81,7 @@ RUN conda install -y -q --name neuro numpy=1.14.1 \
                                      seaborn=0.8.1 \
                                      pytables=3.4.2 \
                                      pandas=0.22.0 \
-                                     nipype=1.0.3 \
+                                     nipype=1.1.1 \
                                      patsy \
     && sync && conda clean -tipsy && sync
 

--- a/fitlins/__about__.py
+++ b/fitlins/__about__.py
@@ -34,19 +34,23 @@ SETUP_REQUIRES = [
 
 REQUIRES = [
     'nibabel>=2.0',
-    'nipype>=1.0.3',
+    'nipype>=1.1.0',
     'seaborn>=0.7.1',
     'numpy>=1.11',
     'nilearn>=0.4.0',
     'pandas>=0.19',
     'tables>=3.2.1',
     'nistats>=0.0.1a',
-    'grabbit>=0.2.1',
-    'pybids>=0.6.3',
+    'grabbit>=0.2.2',
+    'pybids>=0.6.4',
     'jinja2',
 ]
 
 LINKS_REQUIRES = [
+    'git+https://github.com/INCF/pybids.git@'
+    'a86a0e862e93001536bff28e4c2ba35906824d2c#egg=pybids',
+    'git+https://github.com/nistats/nistats.git@'
+    'ce3695e8f34c6f34323766dc96a60a53b69d2729#egg=nistats',
 ]
 
 TESTS_REQUIRES = [

--- a/fitlins/cli/run.py
+++ b/fitlins/cli/run.py
@@ -183,7 +183,7 @@ def run_fitlins(argv=None):
 
     for model in models:
         analysis = ba.Analysis(layout, model=model)
-        report_dicts = parse_directory(deriv_dir, analysis)
+        report_dicts = parse_directory(deriv_dir, opts.work_dir, analysis)
         write_report('unknown', report_dicts, run_context, deriv_dir)
 
     return retcode

--- a/fitlins/data/report.tpl
+++ b/fitlins/data/report.tpl
@@ -10,6 +10,11 @@
 .elem-title {}
 .elem-desc {}
 .elem-filename {}
+.warning {
+    border: 1px solid #ffaaaa;
+    background: #ffe8e8;
+    padding: 0.8em;
+}
 </style>
 </head>
 <body>

--- a/fitlins/data/report.tpl
+++ b/fitlins/data/report.tpl
@@ -23,6 +23,7 @@
     </div>
     <div id="model">
         <h1 class="sub-report-title">Model</h1>
+        {{ warning }}
         {% if design_matrix_svg %}
         <h2>Design matrix</h3>
         <img src="{{ design_matrix_svg }}" />

--- a/fitlins/interfaces/bids.py
+++ b/fitlins/interfaces/bids.py
@@ -205,8 +205,7 @@ class LoadLevel1BIDSModel(SimpleInterface):
             ent_string = '_'.join('{}-{}'.format(key, val)
                                   for key, val in ents.items())
 
-            events_file = os.path.join(runtime.cwd,
-                                       '{}_events.h5'.format(ent_string))
+            events_file = cwd / '{}_events.h5'.format(ent_string)
             paradigm.to_hdf(events_file, key='events')
 
             imputed = []
@@ -243,15 +242,14 @@ class LoadLevel1BIDSModel(SimpleInterface):
                     iflogger.warning('Unexpected NaNs found in confounds; '
                                      'regression may fail.')
 
-                confounds_file = os.path.join(runtime.cwd,
-                                              '{}_confounds.h5'.format(ent_string))
+                confounds_file = cwd / '{}_confounds.h5'.format(ent_string)
                 confounds.to_hdf(confounds_file, key='confounds')
+
             else:
                 confounds_file = None
 
-
-            info['events'] = events_file
-            info['confounds'] = confounds_file
+            info['events'] = str(events_file)
+            info['confounds'] = str(confounds_file)
             info['repetition_time'] = TR
 
             contrasts = block.get_contrasts([contrast['name']
@@ -272,7 +270,7 @@ class LoadLevel1BIDSModel(SimpleInterface):
 
             entities.append(ents)
             session_info.append(info)
-            contrast_info.append(contrasts_file)
+            contrast_info.append(str(contrasts_file))
             warnings.append(str(warning_file))
 
         runtime.analysis = analysis

--- a/fitlins/interfaces/bids.py
+++ b/fitlins/interfaces/bids.py
@@ -404,6 +404,8 @@ class BIDSDataSink(IOBase):
     def _list_outputs(self):
         base_dir = self.inputs.base_directory
 
+        os.makedirs(base_dir, exist_ok=True)
+
         layout = gb.BIDSLayout(base_dir)
         if self.inputs.path_patterns:
             layout.path_patterns[:0] = self.inputs.path_patterns

--- a/fitlins/interfaces/bids.py
+++ b/fitlins/interfaces/bids.py
@@ -1,7 +1,9 @@
 import os
+from pathlib import Path
 from gzip import GzipFile
 import json
 import shutil
+import numpy as np
 import nibabel as nb
 from nipype import logging
 from nipype.utils.filemanip import makedirs, copyfile
@@ -114,6 +116,16 @@ class ModelSpecLoader(SimpleInterface):
         return runtime
 
 
+IMPUTATION_SNIPPET = """
+<div class="warning">
+    The following confounds had NaN values for the first volume: {}.
+    The mean of non-zero values for the remaining entries was imputed.
+    If another strategy is desired, it must be cxplicitly specified in
+    the model.
+</div>
+"""
+
+
 class LoadLevel1BIDSModelInputSpec(BaseInterfaceInputSpec):
     bids_dir = Directory(exists=True,
                          mandatory=True,
@@ -132,8 +144,9 @@ class LoadLevel1BIDSModelInputSpec(BaseInterfaceInputSpec):
 
 class LoadLevel1BIDSModelOutputSpec(TraitedSpec):
     session_info = traits.List(traits.Dict())
-    contrast_info = traits.List(File())
+    contrast_info = traits.List(File)
     entities = traits.List(traits.Dict())
+    warnings = traits.List(File)
 
 
 class LoadLevel1BIDSModel(SimpleInterface):
@@ -141,6 +154,7 @@ class LoadLevel1BIDSModel(SimpleInterface):
     output_spec = LoadLevel1BIDSModelOutputSpec
 
     def _run_interface(self, runtime):
+        cwd = Path(runtime.cwd)
         include = self.inputs.include_pattern
         exclude = self.inputs.exclude_pattern
         if not isdefined(include):
@@ -162,6 +176,7 @@ class LoadLevel1BIDSModel(SimpleInterface):
         entities = []
         session_info = []
         contrast_info = []
+        warnings = []
         for paradigm, _, ents in block.get_design_matrix(
                 block.model['HRF_variables'], mode='sparse', force=True):
             info = {}
@@ -194,6 +209,7 @@ class LoadLevel1BIDSModel(SimpleInterface):
                                        '{}_events.h5'.format(ent_string))
             paradigm.to_hdf(events_file, key='events')
 
+            imputed = []
             if confounds is not None:
                 # Note that FMRIPREP includes CosineXX columns to accompany
                 # t/aCompCor
@@ -221,6 +237,7 @@ class LoadLevel1BIDSModel(SimpleInterface):
 
                         # Impute the mean non-zero, non-NaN value
                         confounds[imputable][0] = np.nanmean(vals[vals != 0])
+                        imputed.append(imputable)
 
                 if np.isnan(confounds.values).any():
                     iflogger.warning('Unexpected NaNs found in confounds; '
@@ -248,15 +265,22 @@ class LoadLevel1BIDSModel(SimpleInterface):
                                           '{}_contrasts.h5'.format(ent_string))
             contrasts.to_hdf(contrasts_file, key='contrasts')
 
+            warning_file = cwd / '{}_warning.html'.format(ent_string)
+            with warning_file.open('w') as fobj:
+                if imputed:
+                    fobj.write(IMPUTATION_SNIPPET.format(', '.join(imputed)))
+
             entities.append(ents)
             session_info.append(info)
             contrast_info.append(contrasts_file)
+            warnings.append(str(warning_file))
 
         runtime.analysis = analysis
 
         self._results['entities'] = entities
         self._results['session_info'] = session_info
         self._results['contrast_info'] = contrast_info
+        self._results['warnings'] = warnings
         return runtime
 
 

--- a/fitlins/interfaces/bids.py
+++ b/fitlins/interfaces/bids.py
@@ -120,7 +120,7 @@ IMPUTATION_SNIPPET = """
 <div class="warning">
     The following confounds had NaN values for the first volume: {}.
     The mean of non-zero values for the remaining entries was imputed.
-    If another strategy is desired, it must be cxplicitly specified in
+    If another strategy is desired, it must be explicitly specified in
     the model.
 </div>
 """

--- a/fitlins/interfaces/bids.py
+++ b/fitlins/interfaces/bids.py
@@ -199,9 +199,21 @@ class LoadLevel1BIDSModel(SimpleInterface):
                 names = [col for col in confounds.columns
                          if col.startswith('NonSteadyStateOutlier') or
                          col in block.model['variables']]
+                confounds = confounds[names]
+                for imputable in ('FramewiseDisplacement',
+                                  'stdDVARS', 'non-stdDVARS',
+                                  'vx-wisestdDVARS'):
+                    if imputable in confounds.columns:
+                        vals = confounds[imputable].values
+                        if not np.isnan(vals).any():
+                            continue
+
+                        # Impute the mean non-zero, non-NaN value
+                        meanval = np.nanmean(vals[vals != 0])
+                        confounds[imputable] = confounds[imputable].fillna(meanval)
                 confounds_file = os.path.join(runtime.cwd,
                                               '{}_confounds.h5'.format(ent_string))
-                confounds[names].fillna(0).to_hdf(confounds_file, key='confounds')
+                confounds.to_hdf(confounds_file, key='confounds')
             else:
                 confounds_file = None
 

--- a/fitlins/interfaces/bids.py
+++ b/fitlins/interfaces/bids.py
@@ -200,17 +200,25 @@ class LoadLevel1BIDSModel(SimpleInterface):
                          if col.startswith('NonSteadyStateOutlier') or
                          col in block.model['variables']]
                 confounds = confounds[names]
+
+                # These confounds are defined pairwise with the current volume
+                # and its predecessor, and thus may be undefined (have value
+                # NaN) at the first volume.
+                # In these cases, we impute the mean non-zero value, for the
+                # expected NaN only.
+                # Any other NaNs must be handled by an explicit transform in
+                # the BIDS model.
                 for imputable in ('FramewiseDisplacement',
                                   'stdDVARS', 'non-stdDVARS',
                                   'vx-wisestdDVARS'):
                     if imputable in confounds.columns:
                         vals = confounds[imputable].values
-                        if not np.isnan(vals).any():
+                        if not np.isnan(vals[0]):
                             continue
 
                         # Impute the mean non-zero, non-NaN value
-                        meanval = np.nanmean(vals[vals != 0])
-                        confounds[imputable] = confounds[imputable].fillna(meanval)
+                        confounds[imputable][0] = np.nanmean(vals[vals != 0])
+
                 confounds_file = os.path.join(runtime.cwd,
                                               '{}_confounds.h5'.format(ent_string))
                 confounds.to_hdf(confounds_file, key='confounds')

--- a/fitlins/interfaces/bids.py
+++ b/fitlins/interfaces/bids.py
@@ -399,7 +399,7 @@ class BIDSDataSink(IOBase):
     input_spec = BIDSDataSinkInputSpec
     output_spec = BIDSDataSinkOutputSpec
 
-    _always_run=True
+    _always_run = True
 
     def _list_outputs(self):
         base_dir = self.inputs.base_directory

--- a/fitlins/interfaces/bids.py
+++ b/fitlins/interfaces/bids.py
@@ -3,6 +3,7 @@ from gzip import GzipFile
 import json
 import shutil
 import nibabel as nb
+from nipype import logging
 from nipype.utils.filemanip import makedirs, copyfile
 from nipype.interfaces.base import (
     BaseInterfaceInputSpec, TraitedSpec, SimpleInterface,
@@ -13,6 +14,8 @@ from nipype.interfaces.io import IOBase
 from bids import grabbids as gb, analysis as ba
 
 from ..utils import snake_to_camel
+
+iflogger = logging.getLogger('nipype.interface')
 
 
 def bids_split_filename(fname):
@@ -218,6 +221,10 @@ class LoadLevel1BIDSModel(SimpleInterface):
 
                         # Impute the mean non-zero, non-NaN value
                         confounds[imputable][0] = np.nanmean(vals[vals != 0])
+
+                if np.isnan(confounds.values).any():
+                    iflogger.warning('Unexpected NaNs found in confounds; '
+                                     'regression may fail.')
 
                 confounds_file = os.path.join(runtime.cwd,
                                               '{}_confounds.h5'.format(ent_string))

--- a/fitlins/interfaces/bids.py
+++ b/fitlins/interfaces/bids.py
@@ -116,7 +116,7 @@ class ModelSpecLoader(SimpleInterface):
         return runtime
 
 
-IMPUTATION_SNIPPET = """
+IMPUTATION_SNIPPET = """\
 <div class="warning">
     The following confounds had NaN values for the first volume: {}.
     The mean of non-zero values for the remaining entries was imputed.

--- a/fitlins/viz/reports.py
+++ b/fitlins/viz/reports.py
@@ -1,7 +1,9 @@
 from os import path as op
+from pathlib import Path
+import grabbit
 import jinja2
 import pkg_resources as pkgr
-from bids import grabbids
+import bids
 
 from ..utils import snake_to_camel
 
@@ -26,10 +28,11 @@ def deroot(val, root):
     return val
 
 
-def parse_directory(deriv_dir, analysis):
-    fl_layout = grabbids.BIDSLayout(
+def parse_directory(deriv_dir, work_dir, analysis):
+    fl_layout = bids.BIDSLayout(
         (deriv_dir, ['bids', 'derivatives',
                      pkgr.resource_filename('fitlins', 'data/fitlins.json')]))
+    wd_layout = bids.BIDSLayout(str(Path(work_dir) / 'reportlets' / 'fitlins'))
     contrast_svgs = fl_layout.get(extensions='.svg', type='contrasts')
 
     analyses = []
@@ -53,6 +56,11 @@ def parse_directory(deriv_dir, analysis):
         if design_matrix:
             job_desc['design_matrix_svg'] = design_matrix[0].filename
 
+        snippet = wd_layout.get(extensions='.html', type='snippet', **ents)
+        if snippet:
+            with open(snippet[0].filename) as fobj:
+                job_desc['warning'] = fobj.read()
+
         contrasts = fl_layout.get(extensions='.png', type='ortho', **ents)
         # TODO: Split contrasts from estimates
         job_desc['contrasts'] = [{'image_file': c.filename,
@@ -66,7 +74,7 @@ def parse_directory(deriv_dir, analysis):
 
 
 def write_report(level, report_dicts, run_context, deriv_dir):
-    fl_layout = grabbids.BIDSLayout(
+    fl_layout = bids.BIDSLayout(
         (deriv_dir, ['bids', 'derivatives',
                      pkgr.resource_filename('fitlins', 'data/fitlins.json')]))
     fl_layout.path_patterns = PATH_PATTERNS

--- a/fitlins/workflows/base.py
+++ b/fitlins/workflows/base.py
@@ -49,6 +49,16 @@ def init_fitlins_wf(bids_dir, preproc_dir, out_dir, space, exclude_pattern=None,
         iterfield=['session_info', 'contrast_info', 'bold_file', 'mask_file'],
         name='flm')
 
+
+    snippet_pattern = '[sub-{subject}/][ses-{session}/][sub-{subject}_]' \
+        '[ses-{session}_]task-{task}_[run-{run}_]snippet.html'
+    ds_model_warnings = pe.MapNode(
+        BIDSDataSink(base_directory=str(Path(base_dir) / 'reportlets'),
+                     path_patterns=snippet_pattern),
+        iterfield=['entities', 'in_file'],
+        run_without_submitting=True,
+        name='ds_model_warning')
+
     contrast_pattern = '[sub-{subject}/][ses-{session}/][sub-{subject}_]' \
         '[ses-{session}_]task-{task}_[run-{run}_]bold[_space-{space}]_' \
         'contrast-{contrast}_{type<effect>}.nii.gz',
@@ -110,6 +120,8 @@ def init_fitlins_wf(bids_dir, preproc_dir, out_dir, space, exclude_pattern=None,
         (loader, getter,  [('entities', 'entities')]),
         (loader, flm, [('session_info', 'session_info'),
                        ('contrast_info', 'contrast_info')]),
+        (loader, ds_model_warnings, [('entities', 'entities'),
+                                     ('warnings', 'in_file')]),
         (getter, flm, [('bold_files', 'bold_file'),
                        ('mask_files', 'mask_file')]),
         (getter, ds_estimate_maps, [('entities', 'fixed_entities')]),

--- a/fitlins/workflows/base.py
+++ b/fitlins/workflows/base.py
@@ -1,3 +1,4 @@
+from pathlib import Path
 from nipype.pipeline import engine as pe
 from ..interfaces.bids import (
     ModelSpecLoader, LoadLevel1BIDSModel, BIDSSelect, BIDSDataSink)
@@ -53,7 +54,7 @@ def init_fitlins_wf(bids_dir, preproc_dir, out_dir, space, exclude_pattern=None,
     snippet_pattern = '[sub-{subject}/][ses-{session}/][sub-{subject}_]' \
         '[ses-{session}_]task-{task}_[run-{run}_]snippet.html'
     ds_model_warnings = pe.MapNode(
-        BIDSDataSink(base_directory=str(Path(base_dir) / 'reportlets'),
+        BIDSDataSink(base_directory=str(Path(base_dir) / 'reportlets' / 'fitlins'),
                      path_patterns=snippet_pattern),
         iterfield=['entities', 'in_file'],
         run_without_submitting=True,

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,9 +1,9 @@
 nibabel>=2.0
-nipype>=1.0.3
+nipype>=1.1.0
 seaborn>=0.7.1
 numpy>=1.11
 nilearn>=0.4.0
 pandas>=0.19
-grabbit>=0.2.1
-pybids>=0.6.3
-git+https://github.com/nistats/nistats.git@ce3695e8f34c6f34323766dc96a60a53b69d2729#egg=nistats-0.0.1b-dev
+grabbit>=0.2.2
+git+https://github.com/INCF/pybids.git@a86a0e862e93001536bff28e4c2ba35906824d2c#egg=pybids
+git+https://github.com/nistats/nistats.git@ce3695e8f34c6f34323766dc96a60a53b69d2729#egg=nistats

